## Importing Libraries & getting the dataset

In [44]:
import numpy as np
import pandas as pd
from pydataset import data
df = data('mtcars')
df.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
Mazda RX4          21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
Mazda RX4 Wag      21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
Datsun 710         22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
Hornet 4 Drive     21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

                   carb  
Mazda RX4             4  
Mazda RX4 Wag         4  
Datsun 710            1  
Hornet 4 Drive        1  
Hornet Sportabout     2

#### Resetting the index as a new column and giving it a col name

In [45]:
df = df.reset_index(names='CarType')
df.head()

CarType   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

### Checking the INFO and understanding the dataset

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CarType  32 non-null     object 
 1   mpg      32 non-null     float64
 2   cyl      32 non-null     int64  
 3   disp     32 non-null     float64
 4   hp       32 non-null     int64  
 5   drat     32 non-null     float64
 6   wt       32 non-null     float64
 7   qsec     32 non-null     float64
 8   vs       32 non-null     int64  
 9   am       32 non-null     int64  
 10  gear     32 non-null     int64  
 11  carb     32 non-null     int64  
dtypes: float64(5), int64(6), object(1)
memory usage: 3.1+ KB


### Checking for NULL values

In [47]:
df.isna().sum()

CarType    0
mpg        0
cyl        0
disp       0
hp         0
drat       0
wt         0
qsec       0
vs         0
am         0
gear       0
carb       0
dtype: int64

In [48]:
### No NULL values in any column - We're good to go

## ML Model - Before Analysis

In [94]:
## Copying the data from df to ml dataframe
ml = df.copy()

### Importing all models, preprocessors, metrics

In [95]:
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, BaggingRegressor, GradientBoostingRegressor
from catboost import CatBoostRegressor
from xgboost import XGBRegressor, XGBRFRegressor
from lightgbm import LGBMRegressor
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error, root_mean_squared_error, mean_absolute_error

In [96]:
ml.head()

CarType   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

### CarType is String --> need to remove it as it does not add value for prediction

In [97]:
ml = ml.drop(columns=['CarType'])
ml.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

### Splitting and Assigning the dataset as x and y

In [98]:
x = ml.drop(columns=['mpg'])
y = df[['mpg']]

In [99]:
## Data check
x.head()

cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4    8  360.0  175  3.15  3.440  17.02   0   0     3     2

In [100]:
y.head()

mpg
0  21.0
1  21.0
2  22.8
3  21.4
4  18.7

In [101]:
xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size= 0.05, random_state=42)

## Linear Regression

In [102]:
lr = LinearRegression()
lr.fit(xtrain, ytrain)
lr_preds = lr.predict(xtest)
lr_r2 = r2_score(ytest, lr_preds)
lr_rmse = np.sqrt(mean_squared_error(ytest, lr_preds))
lr_mae = mean_absolute_error(ytest, lr_preds)

print('r2: ', lr_r2)
print('rmse: ', lr_rmse)
print('mae: ', lr_mae)

r2:  0.9860481215426536
rmse:  0.5492490254374339
mae:  0.3974200930252634


## Good 👍
##### r2   --> 0 to 1, the more it is closer to 1, the more it is performing good. How good the model is able to learn from the data columns and explain the variance
##### rmse --> sudden spike, it says 0.549.. difference between the predicted values
##### mae  --> gets the mean/avg of all the values and returns +/- 0.39 difference between the predicted values and the actual values

In [103]:
ml['mpg'].describe().reset_index().transpose()

0          1         2     3       4     5     6     7
index  count       mean       std   min     25%   50%   75%   max
mpg     32.0  20.090625  6.026948  10.4  15.425  19.2  22.8  33.9

## K-Nearest Neighbour Regressor

In [104]:
knn = KNeighborsRegressor()
knn.fit(xtrain, ytrain)
knn_preds = knn.predict(xtest)
knn_r2 = r2_score(ytest, knn_preds)
knn_rmse = np.sqrt(mean_squared_error(ytest, knn_preds))
knn_mae = mean_absolute_error(ytest, knn_preds)

print('r2: ', knn_r2)
print('rmse: ', knn_rmse)
print('mae: ', knn_mae)

r2:  0.4046016880564228
rmse:  3.588035674293107
mae:  2.7200000000000006


## 👎

## Decision Tree

In [107]:
dt = DecisionTreeRegressor()
dt.fit(xtrain, ytrain)
dt_preds = dt.predict(xtest)
dt_r2 = r2_score(ytest, dt_preds)
dt_rmse = np.sqrt(mean_squared_error(ytest, dt_preds))
dt_mae = mean_absolute_error(ytest, dt_preds)

print('r2: ', dt_r2)
print('rmse: ', dt_rmse)
print('mae: ', dt_mae)

r2:  0.9609203376112845
rmse:  0.9192388155425123
mae:  0.6500000000000004


## 🫲 r2 is good but rmse and mae is bad so - mid

## Random Forest Regressor

In [108]:
rf = RandomForestRegressor()
rf.fit(xtrain, ytrain)
rf_preds = rf.predict(xtest)
rf_r2 = r2_score(ytest, rf_preds)
rf_rmse = np.sqrt(mean_squared_error(ytest, rf_preds))
rf_mae = mean_absolute_error(ytest, rf_preds)

print('r2: ', rf_r2)
print('rmse: ', rf_rmse)
print('mae: ', rf_mae)

r2:  0.8921521563186493
rmse:  1.527069088155479
mae:  1.1940000000000017


C:\Users\Sharan Kalyan\Anaconda3\envs\mons_py313\Lib\site-packages\sklearn\base.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


## 🫲 r2 is good but rmse and mae is bad so - mid

## AdaBoost Regressor

In [117]:
for i in (AdaBoostRegressor(),BaggingRegressor(),GradientBoostingRegressor(),CatBoostRegressor(),XGBRegressor(),XGBRFRegressor(),LGBMRegressor()):
    reg = i
    reg.fit(xtrain, ytrain)
    reg_preds = reg.predict(xtest)
    reg_r2 = r2_score(ytest, reg_preds)
    reg_rmse = np.sqrt(mean_squared_error(ytest, reg_preds))
    reg_mae = mean_absolute_error(ytest, reg_preds)
    print(i)
    print('r2: ', reg_r2)
    print('rmse: ', reg_rmse)
    print('mae: ', reg_mae)
    print('\n')

C:\Users\Sharan Kalyan\Anaconda3\envs\mons_py313\Lib\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Sharan Kalyan\Anaconda3\envs\mons_py313\Lib\site-packages\sklearn\ensemble\_bagging.py:583: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return column_or_1d(y, warn=True)
C:\Users\Sharan Kalyan\Anaconda3\envs\mons_py313\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


AdaBoostRegressor()
r2:  0.9439312318753375
rmse:  1.1010662735619121
mae:  0.7785714285714302


BaggingRegressor()
r2:  0.9526951092611863
rmse:  1.0113604698622543
mae:  0.8249999999999993


GradientBoostingRegressor()
r2:  0.9996936293554736
rmse:  0.08139102690881717
mae:  0.08115303464995982


Learning rate set to 0.023527
0:	learn: 5.7773284	total: 363us	remaining: 364ms
1:	learn: 5.7147921	total: 749us	remaining: 374ms
2:	learn: 5.6627588	total: 991us	remaining: 330ms
3:	learn: 5.6061560	total: 1.41ms	remaining: 350ms
4:	learn: 5.5629266	total: 1.88ms	remaining: 375ms
5:	learn: 5.4945930	total: 2.07ms	remaining: 342ms
6:	learn: 5.4306842	total: 2.39ms	remaining: 339ms
7:	learn: 5.3749244	total: 2.82ms	remaining: 350ms
8:	learn: 5.3184150	total: 3.18ms	remaining: 351ms
9:	learn: 5.2595051	total: 3.54ms	remaining: 351ms
10:	learn: 5.2049842	total: 3.83ms	remaining: 344ms
11:	learn: 5.1558129	total: 4.16ms	remaining: 343ms
12:	learn: 5.1020627	total: 4.46ms	remaining: 338ms
13:	lea

## We have a winner --> Gradient Boost Regressor 
## Great 🤩😮🤯
##### r2   --> 0 to 1, the more it is closer to 1, the more it is performing good. How good the model is able to learn from the data columns and explain the variance
##### rmse --> sudden spike, it says 0.081..(least of all) difference between the predicted values
##### mae  --> gets the mean/avg of all the values and returns +/- 0.0811.. difference between the predicted values and the actual values